File: dotnet/src/MyDotey.SCF/Filter/AbstractValueFilter.cs
using System;

namespace MyDotey.SCF.Filter
{
    /**
     * @author koqizhao
     *
     * May 21, 2018
     * 
     * apply each filter to the value,
     * if null returned by a filter, stop filtering and return null,
     * otherwise, apply next filter to the value
     */
    public abstract class AbstractValueFilter<V> : ValueFilter<V>
    {
        object ValueFilter.Filter(object v)
        {
            return Filter((V)v);
        }

        public abstract V Filter(V t);

        public override string ToString()
        {
            return GetType().Name;
        }
    }
}

File: java/src/main/java/org/mydotey/scf/filter/AbstractValueFilter.java
package org.mydotey.scf.filter;

import java.util.function.Function;

/**
 * @author koqizhao
 *
 * May 22, 2018
 */
public abstract class AbstractValueFilter<T> implements Function<T, T> {

    @Override
    public abstract boolean equals(Object obj);

    @Override
    public String toString() {
        return String.format("%s { }", getClass().getSimpleName());
    }

}
